xds/internal/balancer/edsbalancer/xds_client_wrapper.go

```diff
@@ -185,43 +149,19 @@ func (c *xdsClientWrapper) replaceXDSClient(newClient xdsClientInterface, newBal
 // - if balancer names are the same, do nothing, and return false
 // - if balancer names are different, create new one, and return true
 func (c *xdsClientWrapper) updateXDSClient(config *EDSConfig, attr *attributes.Attributes) (bool, error) {
-	if attr != nil {
-		if clientFromAttr, _ := attr.Value(xdsinternal.XDSClientID).(xdsClientInterface); clientFromAttr != nil {
-			// This will also clear balancerName, to indicate that client is
-			// from attributes.
-			return c.replaceXDSClient(clientFromAttr, ""), nil
-		}
+	if attr == nil {
+		return false, fmt.Errorf("unexported nil attributes, want attributes with xdsClient")
 	}
-
-	clientConfig, err := bootstrapConfigNew()
-	if err != nil {
-		// TODO: propagate this error to ClientConn, and fail RPCs if necessary.
-		clientConfig = &bootstrap.Config{BalancerName: config.BalancerName}
+	clientFromAttr, _ := attr.Value(xdsinternal.XDSClientID).(xdsClientInterface)
```

Comment:
Does it make sense to have a couple of helper functions to add and get the xdsclient from attributes (like we have for other things)?

I'm not sure what would be the best place to put it though, and what return type should the function have? If it returns the concrete type, then it might interfere with overriding from tests. But if we want it to return the interface, that would mean that the `xdsClientInterface` would have to be made an exported type, and that is not ideal as well. Maybe you have some better ideas.

Reply:
I tried adding this. The interface problem is not too bad. We will need to export the interface, and embed it in the tests, because the test xds clients don't implement all the methods.

I also had another idea: since xdsClient will become a singleton, this can simple call xdsclient.New().

I leaf this as is in this PR, and added a TODO.
In the next PR, I will try the singleton idea, and see which is better.